ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/CommandDeduplication.scala

```diff
@@ -3,90 +3,140 @@
+        if (isDuplicate) {
+          StepContinue(transactionEntry)
+        } else {
+          if (commitContext.preExecute) {
+            // The out of time bounds entry is required in the committer, so we set it to the default value as we stop the steps here with the duplicate rejection
+            commitContext.outOfTimeBoundsLogEntry = Some(DamlLogEntry.getDefaultInstance)
           }
+          duplicateRejection(commitContext, transactionEntry)
         }
-        .getOrElse(StepContinue(transactionEntry))
+      }
+
+      private def duplicateRejection(
+          commitContext: CommitContext,
+          transactionEntry: DamlTransactionEntrySummary,
+      )(implicit loggingContext: LoggingContext) = {
+        rejections.reject(
+          DamlTransactionRejectionEntry.newBuilder
+            .setSubmitterInfo(transactionEntry.submitterInfo)
+            // No duplicate rejection is a definite answer as the deduplication entry will eventually expire.
+            .setDefiniteAnswer(false)
+            .setDuplicateCommand(Duplicate.newBuilder.setDetails("")),
+          "the command is a duplicate",
+          commitContext.recordTime,
+        )
+      }
     }
-  }
 
-  def setDeduplicationEntryStep(defaultConfig: Configuration): Step =
+  def setDeduplicationEntryStep(): Step =
     new Step {
       def apply(commitContext: CommitContext, transactionEntry: DamlTransactionEntrySummary)(
           implicit loggingContext: LoggingContext
       ): StepResult[DamlTransactionEntrySummary] = {
-        val (_, config) = getCurrentConfiguration(defaultConfig, commitContext)
-        // Deduplication duration must be explicitly overwritten in a previous step
-        //  (see [[TransactionCommitter.overwriteDeduplicationPeriodWithMaxDuration]]) and set to ``config.maxDeduplicationTime``.
         if (!transactionEntry.submitterInfo.hasDeduplicationDuration) {
           throw Err.InvalidSubmission("Deduplication duration is not set.")
         }
-        val deduplicateUntil = Conversions.buildTimestamp(
-          transactionEntry.submissionTime
-            .add(parseDuration(transactionEntry.submitterInfo.getDeduplicationDuration))
-            .add(config.timeModel.minSkew)
-        )
         val commandDedupBuilder = DamlCommandDedupValue.newBuilder
-          .setDeduplicatedUntil(deduplicateUntil)
-          .setSubmissionTime(Conversions.buildTimestamp(transactionEntry.submissionTime))
+        commitContext.recordTime
+          .map(Conversions.buildTimestamp) match {
+          case Some(recordTime) =>
+            commandDedupBuilder.setRecordTime(recordTime)
+          case None =>
+            commandDedupBuilder.setMaxRecordTime(
+              commitContext.maximumRecordTime
+                .map(Conversions.buildTimestamp)
+                .getOrElse(
+                  throw Err.InternalError("Maximum record time is not set for pre-execution")
+                )
+            )
```

Comment:
I'd find this easier to read if you extracted a variable.

```suggestion
            val maxRecordTime = commitContext.maximumRecordTime.getOrElse(
              throw Err.InternalError("Maximum record time is not set for pre-execution")
            )
            commandDedupBuilder.setMaxRecordTime(
              Conversions.buildTimestamp(maximumRecordTime)
            )
```